### PR TITLE
semaphore_test.cc: add TimedWaitSpuriousWakeupLinux

### PR DESCRIPTION
**Pull request by dconeybe**
Add a unit test for the Semaphore::TimedWait() spurious wakeup fix in https://github.com/firebase/firebase-cpp-sdk/pull/1021.

### CODE REVIEW

File: app/tests/semaphore_test.cc [hunk 1]
```diff
@@ -76,6 +88,43 @@ TEST(SemaphoreTest, TimedWait) {
       0.20 * firebase::internal::kMillisecondsPerSecond);
 }
 
+#if SEMAPHORE_LINUX
```
Comment (jonsimantov) [inline app/tests/semaphore_test.cc:91]: WDYT about just using #ifdef FIREBASE_ANDROID || FIREBASE_LINUX here and above? To me that's clearer, even though it's not the same as the check in semaphore.h.

File: app/tests/semaphore_test.cc [hunk 2]
```diff
@@ -76,6 +88,43 @@ TEST(SemaphoreTest, TimedWait) {
...
+  // Wait for the "signal sending" thread to terminate, since it references
+  // memory on the stack and we can't have it running after this method returns.
+  signal_sending_thread.Join();
+
+  // Unregister the signal handler for SIGUSR1, since it's no longer needed.
+  signal(SIGUSR1, NULL);
+
+  // Make sure that Semaphore::TimedWait() blocked for the entire timeout, and,
+  // specifically, did NOT return early as a result of the SIGUSR1 interruption.
+  ASSERT_LT(labs((finish_ms - start_ms) -
```
Comment (jonsimantov) [inline app/tests/semaphore_test.cc:122]: Can we just use EXPECT_NEAR here?

File: app/tests/semaphore_test.cc [hunk 3]
```diff
@@ -76,6 +88,43 @@ TEST(SemaphoreTest, TimedWait) {
...
+      [](void* arg) {
+        firebase::internal::Sleep(500);
+        pthread_kill(*static_cast<pthread_t*>(arg), SIGUSR1);
+      },
+      &main_thread);
+
+  // Call Semaphore::TimedWait() and keep track of how long it blocks for.
+  firebase::Semaphore sem(0);
+  int64_t start_ms = firebase::internal::GetTimestamp();
+  EXPECT_FALSE(sem.TimedWait(2 * firebase::internal::kMillisecondsPerSecond));
```
Comment (jonsimantov) [inline app/tests/semaphore_test.cc:110]: 2 * firebase::internal::kMillisecondsPerSecond should be in a constant in this function, for clarity

File: app/tests/semaphore_test.cc [hunk 4]
```diff
@@ -76,6 +88,43 @@ TEST(SemaphoreTest, TimedWait) {
       0.20 * firebase::internal::kMillisecondsPerSecond);
 }
 
+#if SEMAPHORE_LINUX
+// Tests that Timed Wait handles spurious wakeup (Linux/Android specific).
+TEST(SemaphoreTest, TimedWaitSpuriousWakeupLinux) {
+  // Register a dummy signal handler for SIGUSR1; otherwise, sending SIGUSR1
+  // later on will crash the application.
+  signal(SIGUSR1, [](int signum) -> void {});
```
Comment (jonsimantov) [inline app/tests/semaphore_test.cc:96]: Does this actually need the trailing return type syntax?

File: app/tests/semaphore_test.cc [hunk 5]
```diff
@@ -76,6 +88,43 @@ TEST(SemaphoreTest, TimedWait) {
...
+  signal_sending_thread.Join();
+
+  // Unregister the signal handler for SIGUSR1, since it's no longer needed.
+  signal(SIGUSR1, NULL);
+
+  // Make sure that Semaphore::TimedWait() blocked for the entire timeout, and,
+  // specifically, did NOT return early as a result of the SIGUSR1 interruption.
+  ASSERT_LT(labs((finish_ms - start_ms) -
+                 (2 * firebase::internal::kMillisecondsPerSecond)),
+            0.20 * firebase::internal::kMillisecondsPerSecond);
```
Comment (jonsimantov) [inline app/tests/semaphore_test.cc:124]: The allowed delta should probably also be in a constant

File: app/tests/semaphore_test.cc (hunk 1 above)
Comment (dconeybe) [reply app/tests/semaphore_test.cc]: Done.

File: app/tests/semaphore_test.cc (hunk 4 above)
Comment (dconeybe) [reply app/tests/semaphore_test.cc]: No it does not. I've removed the trailing return.

File: app/tests/semaphore_test.cc (hunk 3 above)
Comment (dconeybe) [reply app/tests/semaphore_test.cc]: Done.

File: app/tests/semaphore_test.cc (hunk 2 above)
Comment (dconeybe) [reply app/tests/semaphore_test.cc]: Yes, EXPECT_NEAR works perfectly. Done.

File: app/tests/semaphore_test.cc (hunk 5 above)
Comment (dconeybe) [reply app/tests/semaphore_test.cc]: Done.

File: app/tests/semaphore_test.cc [hunk 6]
```diff
@@ -76,6 +84,50 @@ TEST(SemaphoreTest, TimedWait) {
...
+      [](void* arg) {
+        firebase::internal::Sleep(500);
+        pthread_kill(*static_cast<pthread_t*>(arg), SIGUSR1);
+      },
+      &main_thread);
+
+  // Call Semaphore::TimedWait() and keep track of how long it blocks for.
+  firebase::Semaphore sem(0);
+  auto start_ms = firebase::internal::GetTimestamp();
+  const int timed_wait_timeout = 2 * firebase::internal::kMillisecondsPerSecond;
```
Comment (jonsimantov) [inline app/tests/semaphore_test.cc:110]: I think the constants should probably follow the kConstantNaming syntax

**Review by jonsimantov — APPROVED** [review]
Approved with one nit

File: app/tests/semaphore_test.cc (hunk 6 above)
Comment (dconeybe) [reply app/tests/semaphore_test.cc]: Done.

**Comment by dconeybe** [comment]
@jonsimantov Please re-approve, as your previous approval was dismissed as being "stale".

**Pull request merged**